tests/test_mutamorphic.py

```diff
@@ -0,0 +1,144 @@
+import pytest
+from pytest_lazyfixture import lazy_fixture
+import nltk
+from nltk.corpus import wordnet
+import requests
+import random
+import os
+
+from src import preprocess_data, get_data
+
+#TOOD: env variable
+server_url = "http://localhost:8000"
+
+
+def submit_review(review_data: str):
+    """
+    Submits a review to the server for sentiment analysis.
+
+    Args:
+        review_data (str): The review text to be analyzed.
+
+    Returns:
+        int or None: The sentiment value of the review if the request is successful,
+                     or None if an error occurs.
+
+    Raises:
+        requests.exceptions.RequestException: If an error occurs while sending the request.
+    """
+    payload = {
+        "data": review_data
+    }
+    response = requests.post(server_url + "/predict", data=payload, timeout=1.5)
+    if response.status_code == 200:
+        response_data = response.json()
+        sentiment = response_data.get('sentiment', 0)
+        return sentiment == 1
+    else:
+        return None
+
+
+# Generate a mutant by modifying random words in the input review
+def generate_mutant(review, mutate_count):
+    """
+    Generates a mutant version of the input review by applying random mutations to a specified number of words.
+
+    Parameters:
+    review (str): The review text.
+    mutate_count (int): The number of words to be mutated.
+
+    Returns:
+    str: The mutant review text.
+    """
+    words = review.split()
+
+    if mutate_count > len(words):
+        mutate_count = len(words)
+
+    indices = random.sample(range(len(words)), mutate_count)
+    mutated_words = words.copy()
+
+    for idx in indices:
+        word = words[idx]
+        synonyms = []
+
+        for synset in wordnet.synsets(word):
+            for lemma in synset.lemmas():
+                synonym = lemma.name()
+                if synonym != word:
+                    synonyms.append(synonym)
+
+        if synonyms:
+            mutated_words[idx] = random.choice(synonyms)
+
+    mutant_review = ' '.join(mutated_words)
+    return mutant_review
+
+
+# Test the sentiment analysis model with original and mutant reviews
+def get_assertion_model(review, liked, mutate_count):
+    """
+    Test the sentiment analysis model using mutamorphic testing.
+
+    Parameters:
+    review (str): The review text.
+    liked (bool): The expected sentiment label (True for positive, False for negative).
+    mutate_count (int): The number of words to be mutated.
+    """
+    original_sentiment = submit_review(review)
```

Comment:
I think you would have to mock the HTTP request somehow, because I do not think this will work, since it would require network communication between different containers.

Reply:
I am running the model service inside of the workflow so that the request can be made. This means that the tests will be coupled to the service image. Since the requirements are quite confusing, idk how they want us to actually perform the metamorphic testing (since the requirementS make the service and train decoupled and coupled at the same time)